### PR TITLE
feature: access Header & Cookie in UserDefinedContext

### LINKED ISSUE #129: remove `rocket.Cookie` and `rocket.Header`
This is related to #87, the reason is that of `rocket.Cookie` & `rocket.Header` didn't actually a good interface for getting data, so we plan to use **UserDefinedContext** replace them.

### PR DESCRIPTION
**Pull request by dannypsnl**
refactor: replacing by UDC resolved #129

### CODE REVIEW

**Comment by dannypsnl** [comment]
@shana0440 Could you review this PR?

**Comment by dannypsnl** [comment]
I have a concern that `filler` actucally is very rely on `UserContext`'s cache, then why `UserContext` did not handle this functionality?

**Comment by shana0440** [comment]
Yes, `UserContext` should handle filler. But `RouteFiller` also relies on Handler too, if we put filler into `UserContext` then `UserContext` will rely on Handler, that is the thing I want to avoid.
Before we put filler into `UserContext`, we should deal the coupling between filler and handler.

File: handler.go [hunk 1]
```diff
@@ -66,45 +66,32 @@ func (h *handler) getUserContexts(reqURL []string, req *http.Request) ([]reflect
...
-					h.matchedPath,
-				),
-				newQueryFiller(userContext.QueryParams, req.URL.Query()),
-				newJSONFiller(req.Body),
-			)
-			if err != nil {
-				return nil, err
-			}
-			userContexts[i] = context
+		basicChain := newPipeline().
```
Comment (shana0440) [inline handler.go]: I think we should stick in the `generateContext` then we can achieve the same goal without chain.

Comment (dannypsnl) [reply handler.go]: The reason I fall back to a pipeline is that implementation duplicate at two places.

But I have a new idea is to use primitive `append` to combine the whole chain and we use generator function at the end.

**Comment by dannypsnl** [comment]
> Yes, `UserContext` should handle filler. But `RouteFiller` also relies on Handler too, if we put filler into `UserContext` then `UserContext` will rely on Handler, that is the thing I want to avoid.
> Before we put filler into `UserContext`, we should deal the coupling between filler and handler.

I didn't see why filler depends on the handler, it just received the delegate from the handler.

File: filler.go [hunk 2]
```diff
@@ -93,37 +166,27 @@ func (f *formFiller) fill(ctx reflect.Value) error {
...
-func generateContext(userContext *context.UserContext, fillers ...filler) (reflect.Value, error) {
-	context := reflect.New(userContext.ContextType)
-	for _, filler := range fillers {
-		if err := filler.fill(context); err != nil {
-			return context, err
+func (p *pipeline) run(userContext *context.UserContext) (reflect.Value, error) {
+	ctx := reflect.New(userContext.ContextType)
+	for _, filler := range p.fillers {
+		if err := filler.fill(ctx); err != nil {
+			return ctx, err
```
Comment (dannypsnl) [inline filler.go]: In fact, we would drop context if error happened, so I would return `nil, err`

**Comment by shana0440** [comment]
> > Yes, `UserContext` should handle filler. But `RouteFiller` also relies on Handler too, if we put filler into `UserContext` then `UserContext` will rely on Handler, that is the thing I want to avoid.
> > Before we put filler into `UserContext`, we should deal the coupling between filler and handler.
> 
> I didn't see why filler depends on the handler, it just received the delegate from the handler.

https://github.com/dannypsnl/rocket/pull/160/files#diff-282ca1995221da72d2602c8a46e98e9eR45
`routes`, `matchedPathIndex`, `matchedPath` get from the `Handler`, if we put filler into `UserContext` then we will get a method with a lot of arguments. and those arguments are passed from `route.go`, this makes code hard to trace and understand.
So for now, I want to avoid this problem, otherwise, the code will be hard to refactor.

**Comment by dannypsnl** [comment]
> > > Yes, `UserContext` should handle filler. But `RouteFiller` also relies on Handler too, if we put filler into `UserContext` then `UserContext` will rely on Handler, that is the thing I want to avoid.
> > > Before we put filler into `UserContext`, we should deal the coupling between filler and handler.
> > 
> > 
> > I didn't see why filler depends on the handler, it just received the delegate from the handler.
> 
> https://github.com/dannypsnl/rocket/pull/160/files#diff-282ca1995221da72d2602c8a46e98e9eR45
> `routes`, `matchedPathIndex`, `matchedPath` get from the `Handler`, if we put filler into `UserContext` then we will get a method with a lot of arguments. and those arguments are passed from `route.go`, this makes code hard to trace and understand.
> So for now, I want to avoid this problem, otherwise, the code will be hard to refactor.

Yes, then that also make sense for me, I think we should use another PR to focus on refactoring these codes.

**Pull request merged**